zephyr/lib/regions_mm.c

```diff
@@ -1,54 +1,574 @@
 // SPDX-License-Identifier: BSD-3-Clause
 /*
- * Copyright(c) 2022 Intel Corporation. All rights reserved.
+ * Copyright(c) 2023 Intel Corporation. All rights reserved.
  *
  * Author: Jakub Dabek <jakub.dabek@intel.com>
  */
 
 #include <sof/lib/regions_mm.h>
 
-struct virtual_memory_heap
-	vm_heaps[CONFIG_MP_MAX_NUM_CPUS + VIRTUAL_REGION_COUNT];
+/**
+ * @brief Initialize new heap
+ *
+ * Initializes new heap based on region that heap is to be assigned to
+ * and config.
+ * @param cfg Configuration structure with block structure for the heap
+ * @param memory_region_attribute A virtual region identifier
+ * @param span Flag deciding if heap can do contiguous allocation.
+ *
+ * @retval ptr to a new heap if successful.
+ * @retval NULL on creation failure.
+ */
+struct virtual_memory_heap *vmh_init_heap(const struct heap_config *cfg,
+					  int memory_region_attribute,
+					  bool span)
+{
+	uint32_t core_id = cpu_get_id();
+	struct sys_mm_drv_region *virtual_memory_regions =
+	    (struct sys_mm_drv_region *)sys_mm_drv_query_memory_regions();
+
+	/* Check if we haven't created heap for this region already */
+	if (vmh_get_heap_by_attribute(memory_region_attribute, core_id) != NULL)
+		return NULL;
+
+	struct virtual_memory_heap *new_heap =
+	    rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+		    sizeof(struct virtual_memory_heap));
+
+	if (new_heap == NULL)
+		return NULL;
+
+	list_init(&new_heap->node);
+	struct heap_config new_config = {0};
+
+	/* Initialize list that will keep all active allocations*/
+	list_init(&new_heap->active_allocations);
+
+	/* Search for matching attribute so we place heap on right
+	 * virtual region.
+	 * In case of core heaps regions we count theme from 0 to max
+	 * available cores
+	 */
+	if (memory_region_attribute == MEM_REG_ATTR_CORE_HEAP) {
+		new_heap->virtual_region = &virtual_memory_regions[core_id];
+	} else {
+		for (int i = CONFIG_MP_MAX_NUM_CPUS;
+		     i < VIRTUAL_REGION_COUNT; i++) {
```

Comment:
do not declare i inside for()

Reply:
Why ? If I created those variables outside it would only create chaos since I have more loops = more maintanance, more calls.
I really do not see any point to it. Please explain what is the underlying positive for this .

Reply:
It's just the coding style for this project that says you should never declare loop variables in the loop itself, but as individual declarations probably at the beginning of the parent block.

There is also the fact that we sometimes use compilers that are not compatible with that kind of declaration (blame Cadence for that). I forget if the most recent revisions have proper C99 support.

Reply:
> Using compilers that are not compatible seems like a bad idea itself - why use lackluster tool (I know it is a discussion for the ages) . And from performance point of view I would assume that in loop declaration is better.

The incompatible compiler has the ability to emit certain instructions for our specific implementation of Xtensa that the generic GCC compiler cannot. So unfortunately we do have to deal with that.